### PR TITLE
:sparkles: adding sharedv4-svc multi vol test

### PR DESCRIPTION
**Pull request by lsrinivas-pure**
**What this PR does / why we need it**:
It creates a multi shared v4 svc volumes and mounts to a single pod

**Which issue(s) this PR fixes** (optional)
Closes #PTX-3896

**Special notes for your reviewer**:
Results: http://jenkins.pwx.dev.purestorage.com/view/AutomationFest/job/Torpedo/job/tp-scale-sharedv4/10/console

### CODE REVIEW

File: tests/sharedv4/sharedv4_test.go [hunk 1]
```diff
@@ -0,0 +1,269 @@
...
+				}
+			}
+		})
+
+	})
+
+})
+
+// This test performs sharedv4 nfs server pod termination failover use case
+var _ = Describe("{NfsServerKill}", func() {
```
Comment (pureneelesh) [inline tests/sharedv4/sharedv4_test.go:155]: Name is a bit misleading. I thought we are killing the NFS server process on the node. Instead, we are deleting the node itself. Maybe call it "NFSServerNodeDelete"? Also fix the comment.

Does this test work when running in non-aws environment? If not, do we need to skip it where it is not valid?

File: tests/sharedv4/sharedv4_test.go [hunk 2]
```diff
@@ -0,0 +1,269 @@
...
+	var contexts []*scheduler.Context
+
+	It("has to validate that the new pods started successfully after nfs server node is terminated", func() {
+		contexts = make([]*scheduler.Context, 0)
+		nodesToKill := []node.Node{}
+		var err error
+
+		for i := 0; i < Inst().GlobalScaleFactor; i++ {
+			contexts = append(contexts, ScheduleApplications(fmt.Sprintf("nfskill-%d", i))...)
```
Comment (pureneelesh) [inline tests/sharedv4/sharedv4_test.go:164]: nfskill -> nodekill ?

File: tests/sharedv4/sharedv4_test.go [hunk 3]
```diff
@@ -0,0 +1,269 @@
...
+			for _, v := range appVolumes {
+
+				Step("get attached node and stop the instance", func() {
+					currNodes := node.GetStorageDriverNodes()
+					countOfCurrNodes := len(currNodes)
+
+					attachedNode, err := Inst().V.GetNodeForVolume(v, defaultCommandTimeout, defaultCommandRetry)
+					nodesToKill = append(nodesToKill, *attachedNode)
+
+					// Reboot node and check driver status
```
Comment (pureneelesh) [inline tests/sharedv4/sharedv4_test.go:184]: comment is wrong?

File: tests/sharedv4/sharedv4_test.go [hunk 4]
```diff
@@ -0,0 +1,269 @@
...
+					Step(fmt.Sprintf("stop node : %v having volume: %v attached", attachedNode.Name, v.Name), func() {
```
Comment (pureneelesh) [inline tests/sharedv4/sharedv4_test.go:185]: stop node -> delete node?

File: tests/sharedv4/sharedv4_test.go [hunk 5]
```diff
@@ -0,0 +1,269 @@
...
+				Expect(appVolumes).NotTo(BeEmpty())
+			})
+			for _, v := range appVolumes {
+
+				Step("get attached node and stop the instance", func() {
+					currNodes := node.GetStorageDriverNodes()
+					countOfCurrNodes := len(currNodes)
+
+					attachedNode, err := Inst().V.GetNodeForVolume(v, defaultCommandTimeout, defaultCommandRetry)
+					nodesToKill = append(nodesToKill, *attachedNode)
```
Comment (pureneelesh) [inline tests/sharedv4/sharedv4_test.go:182]: Not sure what is the intention here. Why not just pass only the "attachedNode" to KillANodeAndValidate() function? Why do we append the node each time to the list and then pick a random node to kill?

File: tests/sharedv4/sharedv4_test.go [hunk 6]
```diff
@@ -0,0 +1,269 @@
...
+
+						KillANodeAndValidate(nodesToKill)
+
+						Step("validate applications", func() {
+							for _, ctx := range contexts {
+								ValidateContext(ctx)
+							}
+						})
+
+						Step(fmt.Sprintf("validate node: %v is stopped", attachedNode.Name), func() {
```
Comment (pureneelesh) [inline tests/sharedv4/sharedv4_test.go:195]: It seems to me that this will fail if the node randomly picked by KillANode() function is not that same as the attachedNode. No?

Also,
- should this be done before validating the apps?
- use word "deleted" instead of "stopped"? When I see "stopped", I think of the node "down" not "deleted".

File: tests/sharedv4/sharedv4_test.go [hunk 7]
```diff
@@ -0,0 +1,269 @@
...
+							currNodes = node.GetStorageDriverNodes()
+							isNodePresent := false
+							for _, currNode := range currNodes {
+								if currNode.Name == attachedNode.Name {
+									isNodePresent = true
+									break
+								}
+							}
+							Expect(isNodePresent).To(BeFalse())
+						})
```
Comment (pureneelesh) [inline tests/sharedv4/sharedv4_test.go:205]: We could get rid of isNodePresent by calling Fail() on line 200.

File: tests/sharedv4/sharedv4_test.go [hunk 8]
```diff
@@ -0,0 +1,269 @@
...
+
+						Step(fmt.Sprintf("wait to new instance to start scheduler: %s and volume driver: %s",
+							Inst().S.String(), Inst().V.String()), func() {
+							time.Sleep(2 * time.Minute)
+							currNodes = node.GetStorageDriverNodes()
+							Expect(countOfCurrNodes).To(Equal(len(currNodes)))
+							for _, n := range currNodes {
+
+								err = Inst().S.IsNodeReady(n)
+								Expect(err).NotTo(HaveOccurred())
```
Comment (pureneelesh) [inline tests/sharedv4/sharedv4_test.go:215]: How can we be sure that this will not take more than 2 minutes? Ideally, we should check if the node is ready in a loop with a short sleep interval and a long timeout. But I may be misunderstanding what `IsNodeReady()` does.

File: tests/sharedv4/sharedv4_test.go [hunk 9]
```diff
@@ -0,0 +1,269 @@
...
+
+								err = Inst().V.WaitDriverUpOnNode(n, Inst().DriverStartTimeout)
+								Expect(err).NotTo(HaveOccurred())
+							}
+						})
+
+						Step("validate apps", func() {
```
Comment (pureneelesh) [inline tests/sharedv4/sharedv4_test.go:222]: how is this "validate apps" different from the one on line 189? Maybe reword the "Step" description to clarify why we need both.

File: tests/sharedv4/sharedv4_test.go [hunk 10]
```diff
@@ -0,0 +1,269 @@
...
+
+	Step(fmt.Sprintf("Deleting node [%v]", nodeToKill.Name), func() {
+		logrus.Infof("Instance is of %v ", Inst().N.String())
+		err := Inst().N.DeleteNode(nodeToKill, nodeDeleteTimeoutMins)
+		Expect(err).NotTo(HaveOccurred())
+	})
+
+	Step("Wait for 10 min. to node get replaced by autoscalling group", func() {
+		time.Sleep(10 * time.Minute)
+	})
```
Comment (pureneelesh) [inline tests/sharedv4/sharedv4_test.go:258]: Similar question to the above. Hardcoded sleep interval may be prone to intermittent failures or unnecessarily long sleep. Ideally, use a loop with short sleep interval and long timeout.

File: tests/sharedv4/sharedv4_test.go [hunk 11]
```diff
@@ -0,0 +1,269 @@
...
+
+	err := Inst().S.RefreshNodeRegistry()
+	Expect(err).NotTo(HaveOccurred())
+
+	err = Inst().V.RefreshDriverEndpoints()
+	Expect(err).NotTo(HaveOccurred())
+
+	Step(fmt.Sprintf("Validate number of storage nodes after killing node [%v]", nodeToKill.Name), func() {
+		ValidateClusterSize(int64(len(storageDriverNodes)))
```
Comment (pureneelesh) [inline tests/sharedv4/sharedv4_test.go:267]: I don't get this. I thought the parameter to this function is a set of nodes where a sharedv4 service volume is attached. How can we be sure that it is same as the cluster size? Maybe add a function comment to clarify the expectation.

File: deployments/deploy-ssh.sh [hunk 12]
```diff
@@ -174,7 +179,9 @@ if [[ -z "$TEST_SUITE" || "$TEST_SUITE" == "" ]]; then
             "bin/scheduler_upgrade.test",
             "bin/node_decommission.test",
             "bin/license.test",
-            "bin/upgrade_cluster.test",'
+            "bin/upgrade_cluster.test",
+            "bin/upgrade_cluster.test",
```
[the remark this answers is not in the document]
Comment (lsrinivas-pure) [reply deployments/deploy-ssh.sh:183]: fixed

File: drivers/scheduler/k8s/specs/vdbench-sv4-svc-multivol/px-vdbench-app.yml [hunk 13]
```diff
@@ -0,0 +1,40 @@
...
+          {{ if .ClaimsCount }}
+          args: ["Basic", "5400", "datadir", "1" , "{{ .ClaimsCount }}", "$(POD_NAME)"]
+          {{ end }}
+          env:
+            - name: POD_NAME
+              valueFrom:
+                fieldRef:
+                  fieldPath: metadata.name
+          volumeMounts:
+          {{- range $val := Iterate .ClaimsCount }}
```
[the remark this answers is not in the document]
Comment (lsrinivas-pure) [reply drivers/scheduler/k8s/specs/vdbench-sv4-svc-multivol/px-vdbench-app.yml:31]: yes this spec can be applied to non-service sharedv4 also

File: tests/sharedv4/sharedv4_test.go [hunk 14]
```diff
@@ -0,0 +1,269 @@
...
+	nodeDeleteTimeoutMins    = 7 * time.Minute
+)
+
+func TestSharedV4SVC(t *testing.T) {
+	RegisterFailHandler(Fail)
+
+	var specReporters []Reporter
+	junitReporter := reporters.NewJUnitReporter("/testresults/junit_Sharedv4_SVC.xml")
+	specReporters = append(specReporters, junitReporter)
+	RunSpecsWithDefaultAndCustomReporters(t, "Torpedo: Sharedv4_SVC", specReporters)
```
[the remark this answers is not in the document]
Comment (lsrinivas-pure) [reply tests/sharedv4/sharedv4_test.go:33]: Updated with one-liner

File: tests/sharedv4/sharedv4_test.go [hunk 15]
```diff
@@ -0,0 +1,269 @@
...
+		case 2:
+			frequency = 20
+			timeout = 2 * time.Minute
+		case 1:
+			frequency = 10
+			timeout = 1 * time.Minute
+		default:
+			frequency = 10
+			timeout = 1 * time.Minute
+
```
[the remark this answers is not in the document]
Comment (lsrinivas-pure) [reply tests/sharedv4/sharedv4_test.go:81]: fixed

File: tests/sharedv4/sharedv4_test.go [hunk 16]
```diff
@@ -0,0 +1,269 @@
...
+var _ = Describe("{MultiVolumeMounts}", func() {
+
+	It("has to create multiple sharedv4-svc volumes and mount to sing pod", func() {
+		// set frequency mins depending on the chaos level
+		var frequency int
+		var timeout time.Duration
+		switch Inst().ChaosLevel {
+		case 10:
+			frequency = 100
+			timeout = 10 * time.Minute
```
[the remark this answers is not in the document]
Comment (lsrinivas-pure) [reply tests/sharedv4/sharedv4_test.go:50]: I think this varies based on the test. Here we are using frequency to set the number of claims to be created using Choas level which is global and has a helper function

File: tests/sharedv4/sharedv4_test.go [hunk 17]
```diff
@@ -0,0 +1,269 @@
...
+
+			for _, ctx := range contexts {
+				ctx.ReadinessTimeout = timeout
+				ctx.SkipVolumeValidation = false
+				ValidateContext(ctx)
+
+			}
+		})
+
+		Step("get nodes where app is running and restart volume driver", func() {
```
[the remark this answers is not in the document]
Comment (lsrinivas-pure) [reply tests/sharedv4/sharedv4_test.go:118]: fixed

File: tests/sharedv4/sharedv4_test.go [hunk 18]
```diff
@@ -0,0 +1,269 @@
...
+
+					Step(fmt.Sprintf("validate app %s", attachedNode.Name), func() {
+						ValidateContext(ctx)
+					})
+				}
+			}
+		})
+
+	})
+
```
[the remark this answers is not in the document]
Comment (lsrinivas-pure) [reply tests/sharedv4/sharedv4_test.go:151]: fixed

File: tests/sharedv4/sharedv4_test.go (hunk 1 above)
Comment (lsrinivas-pure) [reply tests/sharedv4/sharedv4_test.go:155]: Updated the name

It should work on non-aws setup also

File: tests/sharedv4/sharedv4_test.go (hunk 2 above)
Comment (lsrinivas-pure) [reply tests/sharedv4/sharedv4_test.go:164]: fixed

File: tests/sharedv4/sharedv4_test.go (hunk 3 above)
Comment (lsrinivas-pure) [reply tests/sharedv4/sharedv4_test.go:184]: fixed

File: tests/sharedv4/sharedv4_test.go (hunk 4 above)
Comment (lsrinivas-pure) [reply tests/sharedv4/sharedv4_test.go:185]: fixed

File: tests/sharedv4/sharedv4_test.go (hunk 5 above)
Comment (lsrinivas-pure) [reply tests/sharedv4/sharedv4_test.go:182]: fixed

File: tests/sharedv4/sharedv4_test.go (hunk 6 above)
Comment (lsrinivas-pure) [reply tests/sharedv4/sharedv4_test.go:195]: fixed

File: tests/sharedv4/sharedv4_test.go (hunk 7 above)
Comment (lsrinivas-pure) [reply tests/sharedv4/sharedv4_test.go:205]: fixed

File: tests/sharedv4/sharedv4_test.go (hunk 8 above)
Comment (lsrinivas-pure) [reply tests/sharedv4/sharedv4_test.go:215]: Yes, IsNodeReadyhas that logic. Here we are waiting 2 mins to start the verification

File: tests/sharedv4/sharedv4_test.go (hunk 9 above)
Comment (lsrinivas-pure) [reply tests/sharedv4/sharedv4_test.go:222]: this is a validation of apps after the new instance is ready. Updated the statement accordingly

File: tests/sharedv4/sharedv4_test.go (hunk 10 above)
Comment (lsrinivas-pure) [reply tests/sharedv4/sharedv4_test.go:258]: Update the logic to check and wait

File: tests/sharedv4/sharedv4_test.go (hunk 11 above)
Comment (lsrinivas-pure) [reply tests/sharedv4/sharedv4_test.go:267]: Removed this now as this might be error-prone for aws as new instance might be kickstarted

File: drivers/scheduler/k8s/specs/vdbench-sv4-r2-mvol/px-vdbench-app.yml [hunk 19]
```diff
@@ -0,0 +1,44 @@
...
+      app: vdbench-sharedv4
+  template:
+    metadata:
+      labels:
+        app: vdbench-sharedv4
+    spec:
+      schedulerName: stork
+      containers:
+        - name: vdbench
+          image: srinav/ocp-vdbench:v3
```
[the remark this answers is not in the document]
Comment (lsrinivas-pure) [reply drivers/scheduler/k8s/specs/vdbench-sv4-r2-mvol/px-vdbench-app.yml:19]: pushed the image to portworx repo

File: drivers/scheduler/k8s/specs/vdbench-sv4-svc-multivol/px-vdbench-app.yml [hunk 20]
```diff
@@ -0,0 +1,40 @@
...
+      labels:
+        app: vdbench-sv4-svc-multivol
+    spec:
+      schedulerName: stork
+      containers:
+        - name: vdbench
+          image: srinav/ocp-vdbench:v3
+          imagePullPolicy: IfNotPresent
+          command: ["./bench_runner.sh"]
+          {{ if .ClaimsCount }}
```
[the remark this answers is not in the document]
Comment (lsrinivas-pure) [reply drivers/scheduler/k8s/specs/vdbench-sv4-svc-multivol/px-vdbench-app.yml:22]: fixed it

File: tests/sharedv4/sharedv4_test.go [hunk 21]
```diff
@@ -0,0 +1,269 @@
...
+}
+
+var _ = BeforeSuite(func() {
+	InitInstance()
+})
+
+// This test performs multi volume mounts to a single  deployment
+var _ = Describe("{MultiVolumeMounts}", func() {
+
+	It("has to create multiple sharedv4-svc volumes and mount to sing pod", func() {
```
[the remark this answers is not in the document]
Comment (lsrinivas-pure) [reply tests/sharedv4/sharedv4_test.go:43]: fixed it

File: tests/sharedv4/sharedv4_test.go [hunk 22]
```diff
@@ -0,0 +1,269 @@
...
+			for i := 0; i < Inst().GlobalScaleFactor; i++ {
+				newContexts := ScheduleApplications(taskName)
+				contexts = append(contexts, newContexts...)
+			}
+
+			for _, ctx := range contexts {
+				ctx.ReadinessTimeout = timeout
+				ctx.SkipVolumeValidation = false
+				ValidateContext(ctx)
+
```
[the remark this answers is not in the document]
Comment (lsrinivas-pure) [reply tests/sharedv4/sharedv4_test.go:114]: fixed it

File: tests/sharedv4/sharedv4_test.go [hunk 23]
```diff
@@ -0,0 +1,269 @@
...
+						time.Sleep(20 * time.Second)
+					})
+
+					Step(fmt.Sprintf("validate app %s", attachedNode.Name), func() {
+						ValidateContext(ctx)
+					})
+				}
+			}
+		})
+
```
[the remark this answers is not in the document]
Comment (lsrinivas-pure) [reply tests/sharedv4/sharedv4_test.go:149]: fixed it

File: tests/sharedv4/sharedv4_test.go [hunk 24]
```diff
@@ -0,0 +1,269 @@
...
+)
+
+const (
+	defaultWaitRebootTimeout = 5 * time.Minute
+	defaultCommandRetry      = 5 * time.Second
+	defaultCommandTimeout    = 1 * time.Minute
+	nodeDeleteTimeoutMins    = 7 * time.Minute
+)
+
+func TestSharedV4SVC(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (lsrinivas-pure) [reply tests/sharedv4/sharedv4_test.go:26]: updated the name with Service

**Review by pureneelesh — APPROVED** [review]

File: drivers/scheduler/k8s/specs/vdbench-sv4-svc-multivol/px-vdbench-app.yml [hunk 25]
```diff
@@ -16,12 +16,10 @@ spec:
       schedulerName: stork
       containers:
         - name: vdbench
-          image: srinav/ocp-vdbench:v3
+          image: portworx/vdbench:multivol
```
[the remark this answers is not in the document]
Comment (lsrinivas-pure) [reply drivers/scheduler/k8s/specs/vdbench-sv4-svc-multivol/px-vdbench-app.yml:19]: I took the existing image portworx/vdbench:torpedo and made some changes required to run this test.
Please let me know if I can add some details to it and where to add it

File: tests/sharedv4/sharedv4_test.go [hunk 26]
```diff
@@ -243,27 +233,34 @@ func TestMain(m *testing.M) {
...
 	})
 
-	Step("Wait for 10 min. to node get replaced by autoscalling group", func() {
-		time.Sleep(10 * time.Minute)
+	Step(fmt.Sprintf("Wait for node: %v to be deleted", nodeToKill.Name), func() {
+		maxWait := 10
+	OUTER:
+		for maxWait > 0 {
+			currNodes := node.GetStorageDriverNodes()
+			for _, currNode := range currNodes {
```
[the remark this answers is not in the document]
Comment (lsrinivas-pure) [reply tests/sharedv4/sharedv4_test.go:249]: fixed

File: tests/sharedv4/sharedv4_test.go [hunk 27]
```diff
@@ -243,27 +233,34 @@ func TestMain(m *testing.M) {
 	os.Exit(m.Run())
 }
 
-func KillANodeAndValidate(storageDriverNodes []node.Node) {
-	rand.Seed(time.Now().Unix())
-	nodeToKill := storageDriverNodes[rand.Intn(len(storageDriverNodes))]
+func KillANodeAndValidate(nodeCount int, nodeToKill node.Node) {
```
[the remark this answers is not in the document]
Comment (lsrinivas-pure) [reply tests/sharedv4/sharedv4_test.go:236]: fixed

File: drivers/scheduler/k8s/specs/vdbench-sv4-svc-multivol/px-vdbench-app.yml (hunk 25 above)
[the remark this answers is not in the document]
Comment (pureneelesh) [reply drivers/scheduler/k8s/specs/vdbench-sv4-svc-multivol/px-vdbench-app.yml:19]: It's your call. Maybe what you have done is enough. But think of it this way. When someone wants to make changes to your container 2-3 years down the line, how will they find out what changes you have made? Hopefully, you are still here, but what if you are not? Is there enough info written down at some obvious place? Or do they have to reverse engineer the container?
